crates/typst-cli/src/compile.rs

```diff
@@ -198,6 +198,41 @@ enum ImageExportFormat {
     Svg,
 }
 
+struct OutputTemplate;
+
+impl OutputTemplate {
+    const INDEXABLE: [&'static str; 3] = ["{p}", "{0p}", "{n}"];
+
+    fn has_indexable_template(output: &str) -> bool {
+        Self::INDEXABLE.iter().any(|template| output.contains(template))
+    }
+
+    fn indexable_template() -> [&'static str; 3] {
+        Self::INDEXABLE
+    }
+
+    fn format(output: &str, this_page: usize, total_pages: usize) -> StrResult<String> {
+        // Find the base 10 width of number `i`
+        fn width(i: usize) -> usize {
+            1 + i.checked_ilog10().unwrap_or(0) as usize
+        }
+
+        let other_templates = ["{t}", "{0t}"];
+        Self::INDEXABLE.iter().chain(other_templates.iter()).try_fold(
+            output.to_string(),
+            |out, template| {
+                let replacement = match *template {
+                    "{p}" => format!("{this_page}"),
+                    "{0p}" | "{n}" => format!("{:01$}", this_page, width(total_pages)),
+                    "{t}" | "{0t}" => format!("{total_pages}"),
+                    _ => bail!("unhandled template placeholder {template}"),
```

Comment:
could this be `unreachable!()`? it shouldn't ever reach a user.

Reply:
I assumed unreachable didn't have a message argument (like the C++ unreachable), but it does. I'll change this